tests/unit/test_charm.py

```diff
@@ -26,3 +126,22 @@ def test_update_status_success(self, *_):
 
         self.harness.charm.on.update_status.emit()
         self.assertEqual(self.harness.charm.unit.status, ActiveStatus("slurmrestd available"))
+
+    @patch("pathlib.Path.read_text")
+    @patch("ops.model.Unit.set_workload_version")
+    def test_upgrade_fail(self, *_):
+        self.harness.charm.on.upgrade_charm.emit()
+        self.assertEqual(self.harness.charm.unit.status, BlockedStatus("Error installing slurmrestd"))
```

Comment:
Might not change anything that much, but you can have the patched `pathlib.Path.read_text` return a string value like "v1.0.0" or something similar.